File: finbyzweb/www/my_tasks.js
// Global state object to manage selections
const state = {
    selected_project: null,
    selected_start_date: null,
    all_expanded: true
};

// Fetch data from the server with filters
async function fetchData() {
    try {
        // Prepare filter parameters
        const filters = {};
        
        // Only add filters that have values
        if (state.selected_project) {
            filters.project = state.selected_project;
        }
        
        if (state.selected_start_date) {
            filters.start_date = state.selected_start_date;
        }
        
        
        // Send filters to server-side function
        const response = await frappe.call({
            method: 'finbyzweb.www.my_tasks.get_data',
            args: {
                filters: filters
            }
        });
        
        
        // If we need to populate project filter, do it here
        if (!state.projects_loaded) {
            populateProjectFilter(response.message);
            state.projects_loaded = true;
        }
        
        // Render the data
        renderTaskTree(response.message);

    } catch (error) {
        console.error('Error fetching data:', error);
        frappe.msgprint({
            title: 'Error',
            indicator: 'red',
            message: `Failed to fetch data: ${error.message || 'Unknown error'}`
        });
    }
}

// Populate project filter dropdown with unique projects
function populateProjectFilter(data) {
    const projectSelect = document.getElementById('project-filter');
    if (!projectSelect) return;
    
    // Clear existing options
    projectSelect.innerHTML = '<option value="">All Projects</option>';
    
    // Get unique projects
    const projects = [...new Set(data
        .filter(item => item.is_project === 1)
        .map(item => ({ id: item.project_id, name: item.task })))];
    
    // Add options
    projects.forEach(project => {
        const option = document.createElement('option');
        option.value = project.id;
        option.textContent = project.name;
        projectSelect.appendChild(option);
    });
}

function renderTaskTree(data) {
    const tableBody = document.querySelector('.task-table tbody');
    tableBody.innerHTML = ''; // Clear existing rows

    if (!data || data.length === 0) {
        const emptyRow = document.createElement('tr');
        emptyRow.innerHTML = '<td colspan="9" style="text-align: center;">No tasks found</td>';
        tableBody.appendChild(emptyRow);
        return;
    }

    const taskIdMap = {};
    const projectMap = {};

    // First pass: Build the maps
    data.forEach(item => {
        if (item.is_project === 1) {
            projectMap[item.project_id] = {
                ...item,
                children: []
            };
        } else if (item.task_id) {
            taskIdMap[item.task_id] = {
                ...item,
                children: []
            };
        }
    });

    // Second pass: Build the hierarchy
    data.forEach(item => {
        if (!item.is_project && item.task_id) {
            const task = taskIdMap[item.task_id];

            if (item.parent_task && taskIdMap[item.parent_task]) {
                taskIdMap[item.parent_task].children.push(task);
            } else if (item.project && projectMap[item.project]) {
                projectMap[item.project].children.push(task);
            }
        }
    });

    // Render all projects first
    Object.values(projectMap).forEach(project => {
        renderProjectRow(tableBody, project, 0);
    });

    // Render orphaned tasks (those without a parent or project)
    Object.values(taskIdMap).forEach(task => {
        if ((task.parent_task && taskIdMap[task.parent_task]) ||
            (task.project && projectMap[task.project])) {
            return;
        }

        renderTaskRow(tableBody, task, 0);
    });
    
    // After rendering, attach click handlers explicitly
    attachClickHandlers();
    
    // Set initial expand/collapse state
    updateExpandCollapseState(state.all_expanded);
}

function renderProjectRow(tableBody, project, level) {
    const row = document.createElement('tr');
    row.classList.add('task-row', 'project-row');
    row.dataset.isProject = '1';
    row.dataset.isGroup = '1';
    row.dataset.taskId = project.project_id;
    row.dataset.level = level.toString();

    const hasChildren = project.children && project.children.length > 0;
    let iconClass = hasChildren ? 'fa-chevron-down' : 'fa-circle';
    let iconStyle = hasChildren ? '' : 'font-size: 8px; color: #ccc;';

    const firstCell = document.createElement('td');
    firstCell.className = 'task-name-cell';

    let indentHTML = '';
    for (let i = 0; i < level; i++) {
        indentHTML += `<span class="indent-spacer" style="display: inline-block; width: 20px;"></span>`;
    }

    firstCell.innerHTML = `
        ${indentHTML}
        <span class="icon-container">
            <i class="fa ${iconClass} collapse-icon" style="${iconStyle}" aria-hidden="true"></i>
        </span>
        <span class="task-name">${project.task}</span>
    `;

    row.appendChild(firstCell);

    const cells = [
        project.type || '',
        project.assignee || '',
        project.progress || '',
        project.exp_start_date || '',
        project.exp_end_date || '',
        project.status_show || '',
        project.expected_time || '',
        project.priority || '',
        project.description || ''
    ];

    cells.forEach(cellContent => {
        const cell = document.createElement('td');
        cell.innerHTML = cellContent;
        row.appendChild(cell);
    });

    tableBody.appendChild(row);

    if (hasChildren) {
        project.children.forEach(task => {
            renderTaskRow(tableBody, task, level + 1, project.project_id);
        });
    }
}

function renderTaskRow(tableBody, task, level, parentId) {
    const row = document.createElement('tr');
    row.classList.add('task-row');
    row.dataset.taskId = task.task_id;
    row.dataset.level = level.toString();

    // Set the correct parent ID
    // First check if the task has a parent_task property provided directly from the server
    if (task.parent_task) {
        row.dataset.parent = task.parent_task;
    } 
    // Otherwise, if it was passed a parentId parameter (from a project or a group task), use that
    else if (parentId) {
        row.dataset.parent = parentId;
    }

    const hasChildren = task.children && task.children.length > 0;
    
    if (hasChildren || task.is_group === 1) {
        row.dataset.isGroup = '1';
    }

    let iconClass = (hasChildren || task.is_group === 1) ? 'fa-chevron-down' : 'fa-circle';
    let iconStyle = (hasChildren || task.is_group === 1) ? '' : 'font-size: 8px; color: #ccc;';

    const firstCell = document.createElement('td');
    firstCell.className = 'task-name-cell';

    // Preserve task.indent if available, otherwise use level
    const indentLevel = task.indent !== undefined ? task.indent : level;
    
    let indentHTML = '';
    for (let i = 0; i < indentLevel; i++) {
        indentHTML += `<span class="indent-spacer" style="display: inline-block; width: 20px;"></span>`;
    }

    firstCell.innerHTML = `
        ${indentHTML}
        <span class="icon-container">
            <i class="fa ${iconClass} collapse-icon" style="${iconStyle}" aria-hidden="true"></i>
        </span>
        <span class="task-name">${task.task}</span>
    `;

    row.appendChild(firstCell);

    const cells = [
        task.type || '',
        task.assignee || '',
        task.progress || '',
        task.exp_start_date || '',
        task.exp_end_date || '',
        task.status_show || '',
        task.expected_time || '',
        task.priority || '',
        task.description || ''
    ];

    cells.forEach(cellContent => {
        const cell = document.createElement('td');
        cell.innerHTML = cellContent;
        row.appendChild(cell);
    });

    tableBody.appendChild(row);

    if (hasChildren) {
        task.children.forEach(childTask => {
            renderTaskRow(tableBody, childTask, level + 1, task.task_id);
        });
    }
}

// Improved show/hide functions for child tasks
function showChildTasks(parentId) {
    document.querySelectorAll(`.task-row[data-parent="${parentId}"]`).forEach(row => {
        row.style.display = '';
        
        // Check if this task was expanded before
        const taskId = row.dataset.taskId;
        const icon = row.querySelector('.collapse-icon');
        
        if (icon && icon.classList.contains('fa-chevron-down') && row.dataset.isGroup === '1') {
            showChildTasks(taskId);
        }
    });
}

function hideChildTasks(parentId) {
    document.querySelectorAll(`.task-row[data-parent="${parentId}"]`).forEach(row => {
        row.style.display = 'none';
        
        // Also hide any children of this row
        const taskId = row.dataset.taskId;
        if (taskId) {
            hideChildTasks(taskId);
        }
    });
}

function handleIconClick(iconElement) {
    const row = iconElement.closest('.task-row');
    if (!row) {
        console.error("No parent row found for this icon");
        return;
    }
    
    if (row.dataset.isGroup === '1' || row.dataset.isProject === '1') {
        const taskId = row.dataset.taskId;
        if (!taskId) {
            console.error("No taskId found on this row", row);
            return;
        }
        

        // Debug what we're looking for
        const children = document.querySelectorAll(`.task-row[data-parent="${taskId}"]`);

        // Toggle expand/collapse
        if (iconElement.classList.contains('fa-chevron-down')) {
            // Collapse
            iconElement.classList.remove('fa-chevron-down');
            iconElement.classList.add('fa-chevron-right');
            hideChildTasks(taskId);
        } else {
            // Expand
            iconElement.classList.remove('fa-chevron-right');
            iconElement.classList.add('fa-chevron-down');
            showChildTasks(taskId);
        }
    } else {
        console.log("");
    }
}

// Function to collapse all tasks
function collapseAll() {
    document.querySelectorAll('.task-row[data-is-group="1"], .task-row[data-is-project="1"]').forEach(row => {
        const icon = row.querySelector('.collapse-icon');
        if (icon && icon.classList.contains('fa-chevron-down')) {
            icon.classList.remove('fa-chevron-down');
            icon.classList.add('fa-chevron-right');
            const taskId = row.dataset.taskId;
            if (taskId) {
                hideChildTasks(taskId);
            }
        }
    });
    state.all_expanded = false;
    updateCollapseButtonText();
}

// Function to expand all tasks
function expandAll() {
    document.querySelectorAll('.task-row[data-is-group="1"], .task-row[data-is-project="1"]').forEach(row => {
        const icon = row.querySelector('.collapse-icon');
        if (icon && icon.classList.contains('fa-chevron-right')) {
            icon.classList.remove('fa-chevron-right');
            icon.classList.add('fa-chevron-down');
            const taskId = row.dataset.taskId;
            if (taskId) {
                showChildTasks(taskId);
            }
        }
    });
    state.all_expanded = true;
    updateCollapseButtonText();
}

// Update expand/collapse state based on all_expanded flag
function updateExpandCollapseState(expand) {
    if (expand) {
        expandAll();
    } else {
        collapseAll();
    }
}

// Update the text of the collapse/expand button
function updateCollapseButtonText() {
    const button = document.getElementById('collapse-expand-btn');
    if (button) {
        button.textContent = state.all_expanded ? 'Collapse All' : 'Show All';
    }
}

// Toggle between collapse all and expand all
function toggleCollapseExpand() {
    if (state.all_expanded) {
        collapseAll();
    } else {
        expandAll();
    }
}

// Function to attach direct click handlers to all icons
function attachClickHandlers() {
    document.querySelectorAll('.collapse-icon').forEach(icon => {
        // Remove any existing event listeners
        const newIcon = icon.cloneNode(true);
        icon.parentNode.replaceChild(newIcon, icon);
        
        // Add new event listener
        newIcon.addEventListener('click', function(e) {
            e.stopPropagation(); // Prevent event bubbling
            handleIconClick(this);
        });
    });
    
    // Also add click handlers to icon containers for better click area
    document.querySelectorAll('.icon-container').forEach(container => {
        // Remove any existing event listeners
        const newContainer = container.cloneNode(true);
        container.parentNode.replaceChild(newContainer, container);
        
        // Add new event listener
        newContainer.addEventListener('click', function(e) {
            e.stopPropagation(); // Prevent event bubbling
            const icon = this.querySelector('.collapse-icon');
            if (icon) {
                handleIconClick(icon);
            }
        });
    });
    
    // Add click handler for collapse/expand button
    const collapseExpandBtn = document.getElementById('collapse-expand-btn');
    if (collapseExpandBtn) {
        collapseExpandBtn.addEventListener('click', toggleCollapseExpand);
    }
    
    // Add change handlers for filters
    const projectFilter = document.getElementById('project-filter');
    if (projectFilter) {
        projectFilter.addEventListener('change', function() {
            state.selected_project = this.value;
            fetchData(); // Refetch with new filters
        });
    }
    
    const startDateFilter = document.getElementById('start-date-filter');
    if (startDateFilter) {
        startDateFilter.addEventListener('change', function() {
            state.selected_start_date = this.value;
            fetchData(); // Refetch with new filters
        });
    }
}

// Create filter elements with improved styling
function createFilterElements() {
    const filterContainer = document.createElement('div');
    filterContainer.className = 'task-filters';
    filterContainer.innerHTML = `
        <div class="filter-container">
            <div class="filter-group">
                <label for="project-filter">Project</label>
                <select id="project-filter">
                    <option value="">All Projects</option>
                </select>
            </div>
            
            <div class="filter-group">
                <label for="start-date-filter">Expected Start Date</label>
                <input type="date" id="start-date-filter">
            </div>
            
            <div class="filter-actions">
                <button id="collapse-expand-btn" class="btn btn-default">Collapse All</button>
            </div>
        </div>
    `;
    
    // Find a good place to insert filters (before the task table)
    const taskTable = document.querySelector('.task-table');
    if (taskTable) {
        taskTable.parentNode.insertBefore(filterContainer, taskTable);
    }
}

// Initialize the page
function initializePage() {
    // Load Font Awesome if not already loaded
    if (!document.querySelector('link[href*="font-awesome"]')) {
        const fontAwesome = document.createElement('link');
        fontAwesome.rel = 'stylesheet';
        fontAwesome.href = 'https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css';
        document.head.appendChild(fontAwesome);
    }

    // Add custom styles for the tree view and enhanced filter styling
    const treeStyles = document.createElement('style');
    treeStyles.textContent = `
        /* Task Row Styles */
        .task-row {
            cursor: pointer;
        }

        .task-row[data-is-group="1"] .task-name {
            font-weight: bold;
        }

        .task-row[data-is-project="1"] {
            background-color: #f5f5f5;
        }

        .collapse-icon {
            margin-right: 5px;
            transition: transform 0.2s;
            width: 16px;
            display: inline-block;
            text-align: center;
            vertical-align: middle;
            cursor: pointer;
        }

        .task-name-cell {
            white-space: nowrap;
            padding-left: 5px !important;
        }

        .indent-spacer {
            display: inline-block;
            width: 20px;
        }

        .icon-container {
            display: inline-block;
            width: 16px;
            text-align: center;
            cursor: pointer;
        }

        .task-description-row {
            background-color: #fafafa;
            font-style: italic;
        }
        
        /* Enhanced Filter Styles */
        .filter-container {
            display: flex;
            align-items: flex-end;
            flex-wrap: wrap;
            margin-bottom: 20px;
            padding: 15px;
            background-color: #f8f9fa;
            border-radius: 6px;
            box-shadow: 0 1px 3px rgba(0,0,0,0.1);
        }
        
        .filter-group {
            margin-right: 20px;
            margin-bottom: 5px;
        }
        
        .filter-group label {
            display: block;
            margin-bottom: 6px;
            font-weight: 500;
            color: #505050;
            font-size: 13px;
        }
        
        .filter-group select,
        .filter-group input[type="date"] {
            padding: 6px 10px;
            border-radius: 4px;
            border: 1px solid #ddd;
            min-width: 180px;
            background-color: white;
            font-size: 14px;
        }
        
        .filter-group select:focus,
        .filter-group input[type="date"]:focus {
            border-color: #8aa2ff;
            outline: none;
            box-shadow: 0 0 0 2px rgba(24, 79, 238, 0.1);
        }
        
        .filter-actions {
            margin-left: auto;
            display: flex;
            align-items: center;
        }
        
        .filter-actions button {
            background-color: #fff;
            border: 1px solid #ddd;
            padding: 6px 12px;
            border-radius: 4px;
            cursor: pointer;
            color: #333;
            font-weight: 500;
            transition: all 0.2s;
        }
        
        .filter-actions button:hover {
            background-color: #f0f0f0;
            border-color: #ccc;
        }
    `;
    document.head.appendChild(treeStyles);

    // Create filter elements
    createFilterElements();

    // Fetch and render data
    fetchData();
}

// Event listener for DOMContentLoaded
document.addEventListener('DOMContentLoaded', function() {
    initializePage();
});

File: finbyzweb/www/my_tasks.py
from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import cstr, getdate

def get_context(context):
    if frappe.session.user == "Guest":
        frappe.throw(_("You need to be logged in to access this page"), frappe.PermissionError)

    context.current_user = frappe.get_doc("User", frappe.session.user)
    context.show_sidebar = True

def execute(filters=None):
    data = get_data(filters)
    return data

@frappe.whitelist()
def get_data(filters=None):
    if isinstance(filters, str):
        try:
            # Try to parse if it's a JSON string
            import json
            filters = json.loads(filters)
        except:
            # If parsing fails, create an empty dict
            filters = {}
    # Get projects the user has permission to access
    projects = get_projects(filters)
    tasks = get_tasks(projects, filters)
    
    if not tasks and not projects:
        return []
    
    return prepare_data(projects, tasks)

def get_projects(filters=None):
    # Prepare filter conditions
    filter_conditions = {"name": ["in", frappe.get_all("Project", pluck="name")]}
    
    # Apply additional filters if provided
    if filters:
        # If specific project is requested
        if filters.get("project"):
            filter_conditions["name"] = filters.get("project")
    
    # Get projects the user has permission to access
    return frappe.get_list("Project", 
        fields=["name", "project_name as subject", "status", "priority", 
                "expected_start_date", "expected_end_date", "percent_complete"],
        filters=filter_conditions,
        order_by="name"
    )

def prepare_data(projects, tasks):
    data = []
    project_task_map = {}
    parent_children_map = {}
    
    # Build parent-child relationships map
    for task in tasks:
        project_task_map.setdefault(task.project, []).append(task)
        
        # Handle both direct parent-child and orphaned tasks
        if task.parent_task:
            parent_children_map.setdefault(task.parent_task, []).append(task)
        else:
            # For tasks without parent_task, map them directly under their project
            parent_children_map.setdefault(task.project, []).append(task)

    # First pass - determine which tasks have children
    task_has_children = {}
    for task in tasks:
        # A task has children if it appears as a key in parent_children_map
        task_has_children[task.name] = task.name in parent_children_map
    
    for project in projects:
        project_tasks = project_task_map.get(project.name, [])
        project_progress = calculate_project_progress(project.name)
        project_progress_display = create_progress_display(project_progress)
        
        if project_tasks:
            # Add project header
            project_data = frappe._dict({
                "project_id": project.name,
                "task": cstr(project.subject),
                "progress": project_progress_display,
                "status_show": create_status_display(project.status),
                "status": project.status,
                "expected_time": "",
                "priority": project.priority,
                "description": "",
                "assignee": "",
                "type": "",
                "indent": 0,
                "exp_start_date": project.expected_start_date,
                "exp_end_date": project.expected_end_date,
                "is_group": 1,  # Projects are always groups
                "is_project": 1
            })
            data.append(project_data)
            
            # Process all tasks for this project
            for task in project_tasks:
                # Only process tasks that don't have a parent or whose parent isn't in our task list
                if not task.parent_task or task.parent_task not in {t.name for t in tasks}:
                    add_task_to_data(data, task, parent_children_map, task_has_children, 1, show_progress=True)

    return data

def get_tasks(projects, filters=None):
    project_names = [project.name for project in projects]
    
    # Prepare base filters
    task_filters = {'project': ['in', project_names]}
    
    # Apply additional filters if provided
    if filters:
        # Date range filters for tasks
        if filters.get("start_date"):
            task_filters['exp_start_date'] = ['=', getdate(filters.get("start_date"))]

        task_filters['status'] = ['!=', 'Cancelled']
    # Get all tasks for the projects the user has permission to access
    return frappe.get_all('Task', 
        filters=task_filters,
        fields=['name', 'subject', 'parent_task', 'project', 'status', 'assignee', 
                'priority', 'description', 'exp_start_date', 'exp_end_date', 
                'completed_on', 'expected_time', 'type'],
        order_by='project, parent_task, name'
    )

# The rest of the functions remain unchanged
def add_task_to_data(data, task, parent_children_map, task_has_children, level, show_progress=False):
    # Add the current task
    task_progress = calculate_task_progress(task.name) if show_progress else None
    task_name = '  ' * level + str(task.subject)
    progress_display = create_progress_display(task_progress) if show_progress else ""
    status_display = create_status_display(task.status)
    
    # Set is_group based on whether the task has children
    is_group = 1 if task_has_children.get(task.name, False) else 0

    data.append(frappe._dict({
        "task": task_name,
        "type": task.type,
        "progress": progress_display,
        "assignee": task.assignee,
        "exp_start_date": task.exp_start_date,
        "exp_end_date": task.exp_end_date,
        "status": task.status,
        "status_show": status_display,
        "expected_time": task.expected_time,
        "priority": task.priority,
        "description": task.description,
        "indent": level,
        "is_group": is_group,  # Dynamically set based on children
        "is_project": 0,
        "project": task.project,
        "task_id": task.name,
        "parent_task": task.parent_task,  # Add the parent_task field explicitly
        "completed_on": task.completed_on
    }))

    # Process children if any exist
    children = parent_children_map.get(task.name, [])
    if children:
        # Sort children by name to maintain consistent order
        children.sort(key=lambda x: x.name)
        for child in children:
            add_task_to_data(data, child, parent_children_map, task_has_children, level + 1, show_progress)

def get_progress_color(progress):
    """
    Get the appropriate color based on the progress percentage
    
    Args:
        progress (int): Progress percentage (0-100)
    
    Returns:
        dict: Color styling for the progress indicator
    """
    if progress is None or progress == 0:
        return {
            'color': '#6b7280',  # Neutral gray
            'background-color': '#f3f4f6'
        }
    
    # If progress is less than 100, return orange/red tones
    if progress < 100:
        return {
            'color': '#c53030',  # Dark red
            'background-color': '#fff5f5'  # Light red background
        }
    # If progress is exactly 100, return green
    elif progress == 100:
        return {
            'color': '#286840',  # Dark green
            'background-color': '#eaf5ee'  # Light green background
        }

def get_status_styles(status):
    """Return color mapping for different statuses"""
    colors = {
        "Unplanned": ("#FFF0F0", "#FF0000"),  # Light red bg, red text
        "Overdue": ("#FFF0F0", "#FF4500"),    # Light red bg, orange-red text
        "In-Progress": ("#FFFFF0", "#997A00"), # Light yellow bg, darker yellow text
        "Open": ("#F8F0FF", "#800080"),        # Light purple bg, purple text
        "Pending Review": ("#FFF8F0", "#FFA500"),
        "Working": ("#FFF8F0", "#FFA500"),
        "Completed": ("#F0FFF0", "#008000"),
        "Cancelled": ("#F0F0F0", "#808080"),
        "Change Pending": ("#F0F8FF", "#4169E1"),
        "Scheduled": ("#F0F8FF", "#0000FF"),
        "Planned": ("#FFF0F8", "#FF69B4"),
        "Template": ("#F0F8FF", "#0000FF")
    }
    return colors.get(status, ("#F0F0F0", "#808080"))

def create_status_display(status):
    """
    Create HTML display for status similar to progress_display
    Returns formatted HTML string with appropriate styling
    """
    if not status:
        return ""
    
    bg_color, text_color = get_status_styles(status)
    
    html = f'''
        <div style="
            display: inline-block;
            padding: 2px 8px;
            border-radius: 10px;
            background-color: {bg_color};
            color: {text_color};
            font-size: 12px;
            font-weight: 500;
            white-space: nowrap;
        ">
            {status}
        </div>
    '''
    return html
          
def create_progress_display(progress):
    """
    Create a styled progress display with color
    
    Args:     
        progress (int or None): Progress percentage
    
    Returns:
        str: HTML-formatted progress display with color styling
    """
    if progress is None:
        return ""
    
    # Get color styling
    progress_colors = get_progress_color(progress)
    
    # Create inline style
    progress_style = (
        f"color:{progress_colors['color']};"
        f"background-color:{progress_colors['background-color']};"
        "border-radius:20px;padding:2px 4px;"
    )
    
    # Return styled progress
    return f"<span style='{progress_style}'>{progress}%</span>"
       
def calculate_project_progress(project_name):
    """
    Calculate project progress based on all tasks within the project
    
    Args:
        project_name (str): Name of the project
    
    Returns:
        int: Overall project progress percentage
    """
    # Get all tasks in the project
    project_tasks = frappe.get_all('Task', 
        filters={'project': project_name},
        fields=['name', 'status']
    )
    
    # If no tasks, return None
    if not project_tasks:
        return None
    
    # Count total and completed tasks
    total_tasks = len(project_tasks)
    completed_tasks = sum(1 for task in project_tasks if task.status == 'Completed')
    
    # Calculate and return progress percentage
    return round((completed_tasks / total_tasks) * 100) if total_tasks > 0 else 0
     
def calculate_task_progress(task_name):
    """
    Calculate task progress based on child tasks' completion
    
    Args:
        task_name (str): Name of the task to calculate progress for
    
    Returns:
        int or None: Percentage of completed child tasks, or None if no children
    """
    # Get all child tasks
    child_tasks = frappe.get_all('Task', 
        filters={'parent_task': task_name},
        fields=['name', 'status']
    )
    
    # If no child tasks, return None
    if not child_tasks:
        return None
    
    # Count total and completed child tasks
    total_tasks = len(child_tasks)
    completed_tasks = sum(1 for task in child_tasks if task.status == 'Completed')
    
    # Calculate and return progress percentage
    return round((completed_tasks / total_tasks) * 100) if total_tasks > 0 else 0